protocol/x/ratelimit/keeper/packet.go

```diff
@@ -0,0 +1,181 @@
+	store.Delete(key)
+}
+
+// Sets the sequence number of a packet that was just sent
+func (k Keeper) SetPendingSendPacket(ctx sdk.Context, channelId string, sequence uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.PendingSendPacketPrefix))
+	key := types.GetPendingSendPacketKey(channelId, sequence)
+	store.Set(key, []byte{1}) // Use a single bit to indicate packet is pending.
+}
+
+// Checks whether the packet sequence number is in the store - indicating that it is a pending
+// packet.
+func (k Keeper) HasPendingSendPacket(ctx sdk.Context, channelId string, sequence uint64) bool {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.PendingSendPacketPrefix))
+	key := types.GetPendingSendPacketKey(channelId, sequence)
+
+	return store.Has(key)
+}
+
+// Middleware implementation for OnAckPacket
+// It is called on the sender chain when a relayer relays back the acknowledgement from the receiver chain.
+// On the dYdX chain, this includes the “response” of the receiver chain for outbound transfer from dYdX.
+func (k Keeper) AcknowledgeIBCTransferPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+) error {
+	// Check whether the ack was a success or error
+	ackResponse, err := util.UnpackAcknowledgementResponseForTransfer(ctx, k.Logger(ctx), acknowledgement)
+	if err != nil {
+		return err
+	}
+
+	// Parse the denom, channelId, and amount from the packet
+	packetInfo, err := util.ParsePacketInfo(packet, types.PACKET_SEND)
+	if err != nil {
+		return err
+	}
+
+	// If the ack was successful, remove the pending packet
+	if ackResponse.Status == types.AckResponseStatus_SUCCESS {
+		k.RemovePendingSendPacket(ctx, packetInfo.ChannelID, packet.Sequence)
+		return nil
+	}
+
+	// If the ack failed, undo the change to the capacity
+	k.UndoSendPacket(ctx, packetInfo.ChannelID, packet.Sequence, packetInfo.Denom, packetInfo.Amount)
+	return nil
+}
+
+// Middleware implementation for OnTimeout
+// It is triggered by a relayer with MsgTimeout on the sender chain when timeoutHeight is
+// reached for a sent packet but acknowledgement has not been received. It should therefore
+// revert the capacity change.
+func (k Keeper) TimeoutIBCTransferPacket(ctx sdk.Context, packet channeltypes.Packet) error {
+	packetInfo, err := util.ParsePacketInfo(packet, types.PACKET_SEND)
+	if err != nil {
+		return err
+	}
+
+	k.UndoSendPacket(ctx, packetInfo.ChannelID, packet.Sequence, packetInfo.Denom, packetInfo.Amount)
+	return nil
+}
+
+// If a SendPacket fails or times out, undo the capacity decrease that happened during the send
+// Idempotent - has no effect on a previously removed pending packet.
+func (k Keeper) UndoSendPacket(
+	ctx sdk.Context,
+	channelId string,
+	sequence uint64,
+	denom string,
+	amount *big.Int,
+) {
+	if !k.HasPendingSendPacket(ctx, channelId, sequence) {
+		return
+	}
+	// Undo'ing a withdrawal is equivalent to processing a deposit.
+	k.ProcessDeposit(ctx, denom, amount)
```

Comment:
Does calling this function have any negative downstream effects? For example logging or metrics that indicate a deposit happened?

Wonder if it's better to have an explicit `UndoWithdrawal` function to prevent these issues (now or in the future if code is modified/refactored)

Reply:
Good point, I renamed `ProcessDeposit` to `IncrementCapacitiesForDenom` and wrapped `UndoWithdrawal` around it (and added metrics)